File: blind75/61_serialize_and_deserialize_binary_tree/serialize_and_deserialize_binary_tree.go
package serialize_and_deserialize_binary_tree

import (
	"strconv"
	"strings"
)

// TreeNode represents a node in a binary tree.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Codec handles serialization and deserialization of binary trees.
type Codec struct{}

// Constructor creates a new Codec instance.
func Constructor() Codec {
	return Codec{}
}

// Serialize encodes a tree to a single string using preorder traversal.
func (c *Codec) Serialize(root *TreeNode) string {
	values := []string{}

	var dfs func(*TreeNode)
	dfs = func(node *TreeNode) {
		if node == nil {
			values = append(values, "null")
			return
		}

		values = append(values, strconv.Itoa(node.Val))
		dfs(node.Left)
		dfs(node.Right)
	}

	dfs(root)
	return strings.Join(values, ",")
}

// Deserialize decodes your encoded data to tree.
func (c *Codec) Deserialize(data string) *TreeNode {
	values := strings.Split(data, ",")
	index := 0

	var dfs func() *TreeNode
	dfs = func() *TreeNode {
		if index >= len(values) || values[index] == "null" {
			index++
			return nil
		}

		val, _ := strconv.Atoi(values[index])
		index++

		node := &TreeNode{Val: val}
		node.Left = dfs()
		node.Right = dfs()

		return node
	}

	return dfs()
}

File: blind75/61_serialize_and_deserialize_binary_tree/serialize_and_deserialize_binary_tree.py
from typing import Optional


class TreeNode:
    """Definition for a binary tree node."""

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Codec:
    """Codec to serialize and deserialize binary trees."""

    def serialize(self, root: Optional[TreeNode]) -> str:
        """
        Encodes a tree to a single string using preorder traversal.

        Args:
            root: Root node of the binary tree

        Returns:
            Serialized string representation of the tree
        """
        values = []

        def dfs(node: Optional[TreeNode]) -> None:
            if node is None:
                values.append("null")
                return

            values.append(str(node.val))
            dfs(node.left)
            dfs(node.right)

        dfs(root)
        return ",".join(values)

    def deserialize(self, data: str) -> Optional[TreeNode]:
        """
        Decodes your encoded data to tree.

        Args:
            data: Serialized string representation

        Returns:
            Root node of the reconstructed tree
        """
        values = data.split(",")
        self.index = 0

        def dfs() -> Optional[TreeNode]:
            if self.index >= len(values) or values[self.index] == "null":
                self.index += 1
                return None

            node = TreeNode(int(values[self.index]))
            self.index += 1

            node.left = dfs()
            node.right = dfs()

            return node

        return dfs()
